File: src/main/java/io/github/jonathanpotts/cartographytable/CommandRefreshMapData.java
package io.github.jonathanpotts.cartographytable;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.URL;
import java.nio.file.FileSystem;
import java.nio.file.FileSystems;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import com.google.gson.JsonParser;
import com.google.gson.stream.JsonReader;

import org.bukkit.ChunkSnapshot;
import org.bukkit.Material;
import org.bukkit.World;
import org.bukkit.block.data.BlockData;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.plugin.Plugin;

import io.github.jonathanpotts.cartographytable.models.BlockModel;
import io.github.jonathanpotts.cartographytable.models.ChunkModel;
import io.github.jonathanpotts.cartographytable.models.ServerModel;
import io.github.jonathanpotts.cartographytable.models.VectorXYZ;
import io.github.jonathanpotts.cartographytable.models.VectorXZ;
import io.github.jonathanpotts.cartographytable.models.WorldModel;

public class CommandRefreshMapData implements CommandExecutor {
    private static final short MIN_Y = 0;
    private static final byte MIN_X = 0;
    private static final byte MAX_X = 15;
    private static final byte MIN_Z = 0;
    private static final byte MAX_Z = 15;
    private static final byte MIN_LIGHT_LEVEL = 0;
    private static final byte MAX_LIGHT_LEVEL = 15;
    private static final String VERSION_MANIFEST = "https://launchermeta.mojang.com/mc/game/version_manifest.json";

    private Plugin plugin;
    private boolean isRefreshing = false;
    private Gson gson;

    public CommandRefreshMapData(Plugin plugin) {
        this.plugin = plugin;

        gson = new GsonBuilder()
            .disableHtmlEscaping()
            .create();
    }

    @Override
    public boolean onCommand(CommandSender sender, Command command, String label, String[] args) {
        if (isRefreshing) {
            plugin.getLogger().info("Map data is already being refreshed");
            return true;
        }

        isRefreshing = true;

        try {
            generateServerData();
        } catch (IOException e) {
            plugin.getLogger().severe("Unable to save server data");
            e.printStackTrace();
            isRefreshing = false;
            return true;
        }

        try {
            generateMaterialsData();
        } catch (IOException e) {
            plugin.getLogger().severe("Unable to save materials data");
            e.printStackTrace();
            isRefreshing = false;
            return true;
        }

        for (World world : plugin.getServer().getWorlds()) {
            try {
                processWorld(world);
            }
            catch (IOException e) {
                plugin.getLogger().severe("Unable to process world (" + world.getName() + ")");
                e.printStackTrace();
                isRefreshing = false;
                return true;
            }
        }

        plugin.getLogger().info("Map data has been refreshed");
        isRefreshing = false;
        return true;
    }

    private void generateServerData() throws IOException {
        ServerModel serverModel = new ServerModel();
        serverModel.motd = plugin.getServer().getMotd();
        serverModel.worlds = new ArrayList<>();

        for (World world : plugin.getServer().getWorlds()) {
            WorldModel worldModel = new WorldModel();
            worldModel.name = world.getName();
            worldModel.spawn = new VectorXYZ(world.getSpawnLocation().getBlockX(), world.getSpawnLocation().getBlockY(), world.getSpawnLocation().getBlockZ());

            serverModel.worlds.add(worldModel);
        }

        String json = gson.toJson(serverModel);

        Path path = plugin.getDataFolder().toPath()
        .resolve("wwwroot")
        .resolve("data")
        .resolve("server.json");

        try {
            if (!Files.exists(path.getParent())) {
                Files.createDirectories(path.getParent());
            }

            Files.write(path, json.getBytes());
        } catch (IOException e) {
            throw new IOException("Unable to save server data", e);
        }
    }

    private void generateMaterialsData() throws IOException {
        Map<Integer, String> materialMap = new HashMap<>();

        for (Material material : Material.values()) {
            if (!material.isBlock()) {
                continue;
            }

            materialMap.put(material.ordinal(), material.createBlockData().getAsString(true).split("\\[")[0]);
        }

        String json = gson.toJson(materialMap);

        Path path = plugin.getDataFolder().toPath()
        .resolve("wwwroot")
        .resolve("data")
        .resolve("materials.json");

        if (!Files.exists(path.getParent())) {
            Files.createDirectories(path.getParent());
        }

        Files.write(path, json.getBytes());

        try {
            downloadMaterials();
        } catch (Exception e) {
            throw new IOException(e);
        }
        
    }

    private void downloadMaterials() throws JsonParseException, IOException {
        JsonObject jObject;

        try {
            jObject = (JsonObject)new JsonParser().parse(new JsonReader(new InputStreamReader(new URL(VERSION_MANIFEST).openStream())));
        } catch (Exception e) {
            throw new JsonParseException("Unable to parse Minecraft version manifest", e);
        }

        String releaseManifest = null;

        String latest = jObject.get("latest").getAsJsonObject().get("release").getAsString();
        for (JsonElement version : jObject.get("versions").getAsJsonArray()) {
            JsonObject versionObject = (JsonObject)version;
            if (!versionObject.get("id").getAsString().equals(latest)) {
                continue;
            }

            releaseManifest = versionObject.get("url").getAsString();
            break;
        }

        if (releaseManifest == null) {
            throw new JsonParseException("Unable to find latest release in Minecraft version manifest");
        }

        try {
            jObject = (JsonObject)new JsonParser().parse(new JsonReader(new InputStreamReader(new URL(releaseManifest).openStream())));
        } catch (Exception e) {
            throw new JsonParseException("Unable to parse latest release manifest", e);
        }

        String client = jObject.get("downloads").getAsJsonObject().get("client").getAsJsonObject().get("url").getAsString();

        Path clientTempFile;
        FileSystem fs;

        try {
            clientTempFile = Files.createTempFile("CartographyTable", ".zip");
            InputStream clientStream = new URL(client).openStream();
            Files.copy(clientStream, clientTempFile, StandardCopyOption.REPLACE_EXISTING);
            clientStream.close();

            fs = FileSystems.newFileSystem(clientTempFile, null);
        } catch (Exception e) {
            throw new IOException("Unable to download textures", e);
        }

        Path texturesPath = plugin.getDataFolder().toPath()
            .resolve("wwwroot")
            .resolve("data")
            .resolve("textures")
            .resolve("block");

        if (!Files.exists(texturesPath)) {
            Files.createDirectories(texturesPath);
        }

        Path blockTextures = fs.getPath("assets", "minecraft", "textures", "block");

        for (Path source : Files.walk(blockTextures).collect(Collectors.toList())) {
            Path destination = texturesPath.resolve(blockTextures.relativize(source).toString());
            if (Files.exists(destination) && Files.isDirectory(destination)) {
                continue;
            }
            Files.copy(source, destination, StandardCopyOption.REPLACE_EXISTING);
        }

        Path modelsPath = plugin.getDataFolder().toPath()
        .resolve("wwwroot")
        .resolve("data")
        .resolve("models")
        .resolve("block");

        if (!Files.exists(modelsPath)) {
            Files.createDirectories(modelsPath);
        }

        Path blockModels = fs.getPath("assets", "minecraft", "models", "block");

        for (Path source : Files.walk(blockModels).collect(Collectors.toList())) {
            Path destination = modelsPath.resolve(blockModels.relativize(source).toString());
            if (Files.exists(destination) && Files.isDirectory(destination)) {
                continue;
            }
            Files.copy(source, destination, StandardCopyOption.REPLACE_EXISTING);
        }

        Files.delete(clientTempFile);
    }

    private void processWorld(World world) throws IOException {
        Path folder = world.getWorldFolder().toPath();
        folder = folder.resolve("region");

        if (!Files.exists(folder)) {
            return;
        }

        List<VectorXZ> chunkLocs = new ArrayList<>();

        List<String> files = null;

        // Get list of generated chunks by looking at region data files.
        try {
            files = Files.list(folder)
                .filter(file -> !Files.isDirectory(file))
                .map(Path::getFileName)
                .map(Path::toString)
                .collect(Collectors.toList());
        } catch (IOException e) {
            throw new IOException("Unable to access world data", e);
        }

        for (String file : files) {
            String[] split = file.split("\\.");

            if (!split[0].equals("r") || split.length != 4 || !split[3].equals("mca")) {
                continue;
            }

            int chunkX = Integer.parseInt(split[1]);
            int chunkZ = Integer.parseInt(split[2]);

            chunkLocs.add(new VectorXZ(chunkX, chunkZ));
        }

        for (VectorXZ chunkLoc : chunkLocs) {
            ChunkModel chunkModel = processChunk(world, chunkLoc.x, chunkLoc.z);

            String json = gson.toJson(chunkModel);

            Path path = plugin.getDataFolder().toPath()
                .resolve("wwwroot")
                .resolve("data")
                .resolve(world.getName())
                .resolve(chunkLoc.x + "." + chunkLoc.z + ".json");
    
            try {
                if (!Files.exists(path.getParent())) {
                    Files.createDirectories(path.getParent());
                }
    
                Files.write(path, json.getBytes());
            } catch (IOException e) {
                throw new IOException("Unable to write chunk file", e);
            }
        }
    }

    private ChunkModel processChunk(World world, int x, int z) {
        ChunkSnapshot chunk = world.getChunkAt(x, z).getChunkSnapshot();
        ChunkModel chunkModel = new ChunkModel();
        chunkModel.blocks = new HashMap<>();

        for (short blockY = MIN_Y; blockY < (short)world.getMaxHeight(); blockY++) {
            for (byte blockX = MIN_X; blockX <= MAX_X; blockX++) {
                for (byte blockZ = MIN_Z; blockZ <= MAX_Z; blockZ++) {
                    BlockModel blockModel = processBlock((short)world.getMaxHeight(), chunk, blockX, blockY, blockZ);

                    if (blockModel == null) {
                        continue;
                    }

                    if (chunkModel.blocks.get(blockY) == null) {
                        chunkModel.blocks.put(blockY, new HashMap<>());
                    }

                    if (chunkModel.blocks.get(blockY).get(blockX) == null) {
                        chunkModel.blocks.get(blockY).put(blockX, new HashMap<>());
                    }

                    chunkModel.blocks.get(blockY).get(blockX).put(blockZ, blockModel);
                }
            }
        }

        return chunkModel;
    }

    private BlockModel processBlock(short maxHeight, ChunkSnapshot chunk, byte x, short y, byte z) {
        BlockData block = chunk.getBlockData(x, y, z);

        if (block.getMaterial().isAir()) {
            boolean surroundedByAir = 
                (y > MIN_Y ? chunk.getBlockData(x, y - 1, z).getMaterial().isAir() : true)
                && (y < maxHeight - 1 ? chunk.getBlockData(x, y + 1, z).getMaterial().isAir() : true)
                // Keep edge blocks on X and Z axes in case of needing light information in a different chunk.
                && (x > MIN_X ? chunk.getBlockData(x - 1, y, z).getMaterial().isAir() : false)
                && (x < MAX_X ? chunk.getBlockData(x + 1, y, z).getMaterial().isAir() : false)
                && (z > MIN_Z ? chunk.getBlockData(x, y, z - 1).getMaterial().isAir() : false)
                && (z < MAX_Z ? chunk.getBlockData(x, y, z + 1).getMaterial().isAir() : false);

            if (surroundedByAir) {
                return null;
            }
        }

        byte emittedLight = (byte)chunk.getBlockEmittedLight(x, y, z);
        byte skyLight = (byte)chunk.getBlockSkyLight(x, y, z);

        BlockModel blockModel = new BlockModel();
        blockModel.mat = block.getMaterial().ordinal();

        String blockData = block.getAsString(true);
        int dataStartIndex = blockData.indexOf("[");

        if (dataStartIndex > 0) {
            int dataEndIndex = blockData.indexOf("]", dataStartIndex);

            if (dataEndIndex > dataStartIndex) {
                blockModel.data = blockData.substring(dataStartIndex + 1, dataEndIndex);
            }
        }

        if (emittedLight != (block.getMaterial().isAir() ? MAX_LIGHT_LEVEL : MIN_LIGHT_LEVEL)) {
            if (blockModel.light == null) {
                blockModel.light = new HashMap<>();
            }

            blockModel.light.put(LightType.EMITTED.ordinal(), emittedLight);
        }

        if (skyLight != (block.getMaterial().isAir() ? MAX_LIGHT_LEVEL : MIN_LIGHT_LEVEL)) {
            if (blockModel.light == null) {
                blockModel.light = new HashMap<>();
            }

            blockModel.light.put(LightType.SKY.ordinal(), skyLight);
        }

        return blockModel;
    }
}

File: src/main/kotlin/io/github/jonathanpotts/cartographytable/CommandRefreshMapData.kt
package io.github.jonathanpotts.cartographytable

import io.github.jonathanpotts.cartographytable.models.*
import io.ktor.client.*
import io.ktor.client.engine.cio.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import kotlinx.coroutines.*
import kotlinx.serialization.*
import kotlinx.serialization.json.*
import org.bukkit.Chunk
import org.bukkit.ChunkSnapshot
import org.bukkit.Material
import org.bukkit.World
import org.bukkit.command.Command
import org.bukkit.command.CommandExecutor
import org.bukkit.command.CommandSender
import org.bukkit.plugin.java.JavaPlugin
import java.io.File
import java.io.IOException
import java.nio.file.FileSystems
import java.nio.file.Files
import java.nio.file.StandardCopyOption

/**
 * Executes the "refresh-map-data" command.
 *
 * @property plugin The server plugin associated to this command.
 * @constructor Creates an instance of the command executor.
 */
class CommandRefreshMapData(private val plugin: JavaPlugin) : CommandExecutor {
    /**
     * Status of the command execution.
     */
    private var isExecuting = false

    /**
     * Folder containing the web app.
     */
    private val webFolder = plugin.dataFolder.resolve("web")

    /**
     * Folder containing data for the web app.
     */
    private val webDataFolder = webFolder.resolve("data")

    override fun onCommand(
        sender: CommandSender,
        cmd: Command,
        label: String,
        args: Array<String>
    ): Boolean {
        if (isExecuting) {
            plugin.logger.info("Map data is already being refreshed")
            return true
        }

        isExecuting = true

        GlobalScope.launch {
            generateServerData()
            generateMaterialData()
            downloadMaterialTexturesAndModels()
            processWorlds()

            isExecuting = false
            SpigotHelpers.runOnServerThread(plugin) {
                plugin.logger.info("Map data has been refreshed")
            }
        }

        return true
    }

    /**
     * Generates server data and saves it.
     */
    private suspend fun generateServerData() {
        val serverModel = ServerModel()

        SpigotHelpers.runOnServerThread(plugin) {
            serverModel.motd = plugin.server.motd

            for (world in plugin.server.worlds) {
                val worldModel = WorldModel()
                worldModel.name = world.name
                worldModel.spawn = VectorXYZ(
                    world.spawnLocation.blockX,
                    world.spawnLocation.blockY,
                    world.spawnLocation.blockZ
                )

                serverModel.worlds.add(worldModel)
            }
        }

        val serverJson = Json.encodeToString(serverModel)

        @Suppress("BlockingMethodInNonBlockingContext")
        withContext(Dispatchers.IO) {
            webDataFolder.mkdirs()
            val licenseFile = webDataFolder.resolve("LICENSE.txt")
            licenseFile.createNewFile()
            licenseFile.writeText(Constants.MINECRAFT_LICENSE_NOTICE)

            val serverFile = webDataFolder.resolve("server.json")
            serverFile.createNewFile()
            serverFile.writeText(serverJson)
        }
    }

    /**
     * Generates material data and saves it.
     */
    private suspend fun generateMaterialData() {
        val materials = SpigotHelpers.runOnServerThread(plugin) {
            Material.values().filter { it.isBlock }
                .associate {
                    Pair(
                        it.ordinal,
                        it.createBlockData()
                            .asString
                            .split("[")
                            .first()
                    )
                }
        }

        val materialsJson = Json.encodeToString(materials)

        @Suppress("BlockingMethodInNonBlockingContext")
        withContext(Dispatchers.IO) {
            webDataFolder.mkdirs()
            val materialsFile = webDataFolder.resolve("materials.json")
            materialsFile.createNewFile()
            materialsFile.writeText(materialsJson)
        }
    }

    /**
     * Downloads material textures and models.
     */
    private suspend fun downloadMaterialTexturesAndModels() {
        val client = HttpClient(CIO)

        val launcherResponse: HttpResponse = client.get(Constants.LAUNCHER_VERSION_MANIFEST)
        val launcherResponseObject = Json.parseToJsonElement(launcherResponse.readText()).jsonObject
        val latestRelease = launcherResponseObject["latest"]?.jsonObject?.get("release")?.jsonPrimitive?.contentOrNull
            ?: throw IOException("Unable to get version for latest release from Minecraft version manifest")
        val latestReleaseManifest = launcherResponseObject["versions"]?.jsonArray?.first {
            it.jsonObject["id"]?.jsonPrimitive?.contentOrNull.equals(latestRelease)
        }?.jsonObject?.get("url")?.jsonPrimitive?.contentOrNull
            ?: throw IOException("Unable to get URL for latest release manifest from Minecraft version manifest")

        val releaseResponse: HttpResponse = client.get(latestReleaseManifest)
        val releaseResponseObject = Json.parseToJsonElement(releaseResponse.readText()).jsonObject
        val mcClient =
            releaseResponseObject["downloads"]?.jsonObject?.get("client")?.jsonObject?.get("url")?.jsonPrimitive?.contentOrNull
                ?: throw IOException("Unable to get URL for client from latest release manifest")

        val clientResponse: HttpResponse = client.get(mcClient)

        @Suppress("BlockingMethodInNonBlockingContext")
        withContext(Dispatchers.IO) {
            val clientFile = File.createTempFile("CartographyTable", ".zip")
            clientFile.writeBytes(clientResponse.readBytes())

            val fs = FileSystems.newFileSystem(clientFile.toPath(), null)

            val blockTexturesFolder = webDataFolder
                .resolve("textures")
                .resolve("block")
            blockTexturesFolder.mkdirs()

            val zipBlockTexturesFolder = fs.rootDirectories.first()
                .resolve("assets")
                .resolve("minecraft")
                .resolve("textures")
                .resolve("block")

            Files.walk(zipBlockTexturesFolder).forEach {
                val relativePath = zipBlockTexturesFolder.relativize(it)
                val destination = blockTexturesFolder.resolve(relativePath.toString())
                if (destination.isDirectory) {
                    destination.mkdirs()
                    return@forEach
                }

                Files.copy(it, destination.toPath(), StandardCopyOption.REPLACE_EXISTING)
            }

            val blockModelsFolder = webDataFolder
                .resolve("models")
                .resolve("block")
            blockModelsFolder.mkdirs()

            val zipBlockModelsFolder = fs.rootDirectories.first()
                .resolve("assets")
                .resolve("minecraft")
                .resolve("models")
                .resolve("block")

            Files.walk(zipBlockModelsFolder).forEach {
                val relativePath = zipBlockModelsFolder.relativize(it)
                val destination = blockModelsFolder.resolve(relativePath.toString())
                if (destination.isDirectory) {
                    destination.mkdirs()
                    return@forEach
                }

                Files.copy(it, destination.toPath(), StandardCopyOption.REPLACE_EXISTING)
            }

            clientFile.delete()
        }
    }

    /**
     * Processes worlds and saves data.
     */
    private suspend fun processWorlds() {
        val worlds = SpigotHelpers.runOnServerThread(plugin) { plugin.server.worlds }
        for (world in worlds) {
            processWorld(world)
        }
    }

    /**
     * Processes a world and saves data.
     *
     * @param world The world to process.
     */
    private suspend fun processWorld(world: World) {
        val worldFolder = SpigotHelpers.runOnServerThread(plugin) { world.worldFolder }
        val regionFolder = worldFolder.resolve("region")
        if (!regionFolder.exists()) {
            return
        }

        val chunkFiles = withContext(Dispatchers.IO) {
            regionFolder.listFiles()?.filter {
                !it.isDirectory && it.name.startsWith("r.") && it.name.split(".").count() == 4 && it.extension == "mca"
            }
        }

        val chunkCoords = chunkFiles?.map {
            val splitName = it.name.split(".")
            VectorXZ(splitName[1].toInt(), splitName[2].toInt())
        }

        chunkCoords?.forEach {
            val chunk = SpigotHelpers.runOnServerThread(plugin) {
                world.getChunkAt(it.x, it.z)
            }

            processChunk(chunk)
        }
    }

    /**
     * Processes a chunk and saves data.
     *
     * @param chunk Chunk to process.
     */
    private suspend fun processChunk(chunk: Chunk) {
        val snapshot = chunk.chunkSnapshot

        val chunkModel = SpigotHelpers.runOnServerThread(plugin) {
            val chunkModel = ChunkModel()

            for (y in Constants.MIN_BLOCK_Y until chunk.world.maxHeight)
                for (x in Constants.MIN_X_FOR_BLOCK_IN_CHUNK..Constants.MAX_X_FOR_BLOCK_IN_CHUNK)
                    for (z in Constants.MIN_Z_FOR_BLOCK_IN_CHUNK..Constants.MAX_Z_FOR_BLOCK_IN_CHUNK) {
                        val blockModel = processBlock(chunk.world.maxHeight, snapshot, x, y, z) ?: continue

                        if (chunkModel.blocks[y] == null) {
                            chunkModel.blocks[y] = mutableMapOf()
                        }

                        if (chunkModel.blocks[y]!![x] == null) {
                            chunkModel.blocks[y]!![x] = mutableMapOf()
                        }

                        chunkModel.blocks[y]!![x]!![z] = blockModel
                    }

            return@runOnServerThread chunkModel
        }

        val chunkJson = Json.encodeToString(chunkModel)

        @Suppress("BlockingMethodInNonBlockingContext")
        withContext(Dispatchers.IO) {
            val worldFolder = webDataFolder
                .resolve(chunk.world.name)
            worldFolder.mkdirs()

            val chunkFile = worldFolder
                .resolve("${chunk.x}.${chunk.z}.json")
            chunkFile.createNewFile()
            chunkFile.writeText(chunkJson)
        }
    }

    /**
     * Processes a block.
     *
     * @param maxHeight Maximum height of the world.
     * @param chunkSnapshot Snapshot of the chunk containing the block.
     * @param x X coordinate of the block in the chunk.
     * @param y Y coordinate of the block in the chunk.
     * @param z Z coordinate of the block in the chunk.
     */
    private fun processBlock(maxHeight: Int, chunkSnapshot: ChunkSnapshot, x: Int, y: Int, z: Int): BlockModel? {
        val blockData = chunkSnapshot.getBlockData(x, y, z)
        if (blockData.material.isAir) {
            var surroundedByAir = true

            if (y > Constants.MIN_BLOCK_Y) {
                surroundedByAir = surroundedByAir && chunkSnapshot.getBlockData(x, y - 1, z).material.isAir
            }

            if (y < maxHeight - 1) {
                surroundedByAir = surroundedByAir && chunkSnapshot.getBlockData(x, y + 1, z).material.isAir
            }

            // Keep edge blocks on X and Z axes in case of needing light information in a different chunk.

            surroundedByAir =
                if (x > Constants.MIN_X_FOR_BLOCK_IN_CHUNK) {
                    surroundedByAir && chunkSnapshot.getBlockData(x - 1, 0, 0).material.isAir
                } else {
                    false
                }

            surroundedByAir =
                if (x < Constants.MIN_X_FOR_BLOCK_IN_CHUNK) {
                    surroundedByAir && chunkSnapshot.getBlockData(x + 1, 0, 0).material.isAir
                } else {
                    false
                }

            surroundedByAir =
                if (z > Constants.MIN_Z_FOR_BLOCK_IN_CHUNK) {
                    surroundedByAir && chunkSnapshot.getBlockData(0, 0, -1).material.isAir
                } else {
                    false
                }

            surroundedByAir =
                if (z < Constants.MIN_Z_FOR_BLOCK_IN_CHUNK) {
                    surroundedByAir && chunkSnapshot.getBlockData(0, 0, 1).material.isAir
                } else {
                    false
                }

            if (surroundedByAir) {
                // If an air block is surrounded by air, the lighting data for that block is not needed.
                return null
            }
        }

        val blockModel = BlockModel()
        blockModel.mat = blockData.material.ordinal

        val data = blockData.asString.substringAfter("[", "").substringBeforeLast("]", "")

        if (data.isNotBlank()) {
            blockModel.data = data
        }

        val defaultLightValue =
            if (blockData.material == Material.AIR) {
                Constants.MAX_LIGHT_LEVEL
            } else {
                Constants.MIN_LIGHT_LEVEL
            }

        val skyLight = chunkSnapshot.getBlockSkyLight(x, y, z)
        val emittedLight = chunkSnapshot.getBlockEmittedLight(x, y, z)

        if (skyLight != defaultLightValue) {
            blockModel.light[LightType.SKY.ordinal] = skyLight
        }

        if (emittedLight != defaultLightValue) {
            blockModel.light[LightType.EMITTED.ordinal] = emittedLight
        }

        return blockModel
    }
}
